### PR TITLE
metronome: add independent volume configuration

### LINKED ISSUE #10347: Metronome Independent Volume Level
**Can't hear volume of metronome when game sounds are on**
When doing Pvm it is often useful to have game sounds on to hear an NPC attack. But it's also useful to hear the metronome. Right now you can barely hear the metronome when you turn on game sounds.

**A volume control for the metronome independent from game sounds**
Add a volume control within the plugin settings that allow the control of the metronome volume, without changing the in-game sounds.

**Right now this is only possible by multi logging with 2 accounts in the same world**
By multi logging into the same world you can still achieve the same result. The weird part is that runelite used to have independent control for the metronome but it was recently removed?

### PR DESCRIPTION
**Pull request by machkr**
Closes #10347. Without setting and resetting the client sound effect volume as implemented, the sound effect is only played at the configured volume when sound effects are muted in-game.

### CODE REVIEW

**Comment by Adam-** [comment]
This seems fine. It may be worth writing a comment in the code explaining why the volume has to be set and reset to avoid getting around muted game sounds.

**Comment by machkr** [comment]
> This seems fine. It may be worth writing a comment in the code explaining why the volume has to be set and reset to avoid getting around muted game sounds.

Added comment

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 1]
```diff
@@ -36,21 +37,37 @@
...
 		return 1;
 	}
 
+	@ConfigItem(
+		keyName = "tickVolume",
+		name = "Tick Volume",
+		description = "Configures the volume of the tick sound",
+		position = 1
+	)
+	default MetronomeVolume tickVolume() { return MetronomeVolume.MEDIUM_HIGH; }
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:53]: [The project's code conventions are to place curly braces on newlines.](https://github.com/runelite/runelite/wiki/Code-Conventions)
```suggestion
	default MetronomeVolume tickVolume()
	{
		return MetronomeVolume.MEDIUM_HIGH;
	}
```

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePlugin.java [hunk 2]
```diff
@@ -69,14 +68,25 @@ public void onGameTick(GameTick tick)
...
 		{
+			// As playSoundEffect only uses the provided volume when in-game sound effects are muted, the sound effect
+			// volume needs to be set to the desired volume then reset after to what it was initially configured to.
+			int previousVolume = client.getSoundEffectVolume();
+
 			if (config.enableTock() && shouldTock)
 			{
-				client.playSoundEffect(SoundEffectID.GE_DECREMENT_PLOP, SoundEffectVolume.MEDIUM_HIGH);
+				int tockVolume = config.tockVolume().getVolume();
+				client.setSoundEffectVolume(tockVolume);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePlugin.java:78]: Why is `setSoundEffectVolume` used along with `playSoundEffect(int id, int volume)`? The documentation for the latter method explicitly states:

> Plays a sound effect, even if the player's sound effect volume is muted.

**Comment by Matsyir** [comment]
I would prefer if the config was a simple int with a range of 0-127, which is what SoundEffectVolume is, a few int constants for each 25% volume tick. Just to have that extra volume precision.
The fix for consistent volume (unrelated to in-game volume) will be really nice, regardless - happy to see this.

**Comment by machkr** [comment]
Tick/tock volume input widgets have been switched from enum dropdowns to int spinners; values of 0 will disable the respective sound. This change removes the need for a tock toggle.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 3]
```diff
@@ -28,29 +28,49 @@
...
+	)
 	@ConfigItem(
-		keyName = "enableTock",
-		name = "Enable tock (alternating) sound",
-		description = "Toggles whether to play two alternating sounds",
-		position = 3
+		keyName = "tockVolume",
+		name = "Tock volume",
+		description = "Configures the volume of the tock sound. A value of 0 will disable tock sounds.",
+		position = 2
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:70]: Positions can be removed.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomeVolume.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ */
+package net.runelite.client.plugins.metronome;
+
+import net.runelite.api.SoundEffectVolume;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
+public enum MetronomeVolume
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomeVolume.java:33]: This is unused?

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 5]
```diff
@@ -28,29 +28,49 @@
 import net.runelite.client.config.Config;
 import net.runelite.client.config.ConfigGroup;
 import net.runelite.client.config.ConfigItem;
+import net.runelite.client.config.Range;
 
 @ConfigGroup("metronome")
 public interface MetronomePluginConfiguration extends Config
 {
+	int VOLUME_MAX = 127;
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:36]:
```suggestion
	private static final int VOLUME_MAX = 127;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java (hunk 3 above)
Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java]: Completed.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomeVolume.java (hunk 4 above)
Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomeVolume.java]: Removed.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java (hunk 5 above)
Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:36]: Interface fields are always "public", "static" and "final". No change necessary.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 6]
```diff
@@ -28,29 +28,46 @@
...
+		max = VOLUME_MAX
+	)
+	@ConfigItem(
+		keyName = "tickVolume",
+		name = "Tick volume",
+		description = "Configures the volume of the tick sound. A value of 0 will disable tick sounds."
+	)
+	default int tickVolume()
+	{
+		return 96;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:58]:
```suggestion
		return SoundEffectVolume.MEDIUM_HIGH;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 7]
```diff
@@ -28,29 +28,46 @@
 import net.runelite.client.config.Config;
 import net.runelite.client.config.ConfigGroup;
 import net.runelite.client.config.ConfigItem;
+import net.runelite.client.config.Range;
 
 @ConfigGroup("metronome")
 public interface MetronomePluginConfiguration extends Config
 {
+	int VOLUME_MAX = 127;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:36]:
```suggestion
	int VOLUME_MAX = SoundEffectVolume.HIGH;
```

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java [hunk 8]
```diff
@@ -28,29 +28,46 @@
...
-		position = 3
+		keyName = "tockVolume",
+		name = "Tock volume",
+		description = "Configures the volume of the tock sound. A value of 0 will disable tock sounds."
 	)
-	default boolean enableTock()
+	default int tockVolume()
 	{
-		return false;
+		return 0;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java:71]:
```suggestion
		return SoundEffectVolume.MUTED;
```

Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java]: Complete.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java (hunk 7 above)
Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java]: Complete.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java (hunk 6 above)
Comment (machkr) [reply runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePluginConfiguration.java]: Complete.

File: runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePlugin.java [hunk 9]
```diff
@@ -69,14 +68,23 @@ public void onGameTick(GameTick tick)
 
 		if (++tickCounter % config.tickCount() == 0)
 		{
-			if (config.enableTock() && shouldTock)
+			// As playSoundEffect only uses the volume argument when the in-game volume is muted, sound effect volume
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/metronome/MetronomePlugin.java:71]: do you mean not muted?

**Pull request merged**